docs/reference/aggregations/bucket/variablewidthhistogram-aggregation.asciidoc

```diff
@@ -0,0 +1,85 @@
+[[search-aggregations-bucket-variablewidthhistogram-aggregation]]
+=== Variable Width Histogram Aggregation
+
+This is a multi-bucket aggregation similar to <<search-aggregations-bucket-histogram-aggregation>>.
+However, the width of each bucket is not specified. Rather, a target number of buckets is provided and bucket intervals
+are dynamically determined based on the document distribution. This is done using a simple one-pass document clustering algorithm
+that aims to obtain low distances between bucket centroids. Unlike other multi-bucket aggregations, the intervals will not
+necessarily have a uniform width.
+
+TIP: The number of buckets returned will always be less than or equal to the target number.
+
+Requesting a target of 3 buckets.
+
+[source,console]
+--------------------------------------------------
+POST /sales/_search?size=0
+{
+    "aggs" : {
+        "prices" : {
+            "variable_width_histogram" : {
+                "field" : "price",
+                "buckets" : 2
+            }
+        }
+    }
+}
+--------------------------------------------------
+// TEST[setup:sales]
+
+Response:
+
+[source,console-result]
+--------------------------------------------------
+{
+    ...
+    "aggregations": {
+        "prices" : {
+            "buckets": [
+                {
+                    "key": 10.0,
+                    "doc_count": 2
+                },
+                {
+                    "key": 150.0,
+                    "doc_count": 5
+                }
+            ]
+        }
+    }
+}
+--------------------------------------------------
+// TESTRESPONSE[s/\.\.\./"took": $body.took,"timed_out": false,"_shards": $body._shards,"hits": $body.hits,/]
+
+==== Shard size
+The `shard_size` parameter specifies the number of buckets that the coordinating node will request from each shard.
+A higher `shard_size` leads each shard to produce smaller buckets. This reduce the likelihood of buckets overlapping
+after the reduction step. Increasing the `shard_size` will improve the accuracy of the histogram, but it will
+also make it more expensive to compute the final result because bigger priority queues will have to be managed on a
+shard level, and the data transfers between the nodes and the client will be larger.
+
+TIP: Parameters `buckets`, `shard_size`, and `cache_limit` are optional. By default, `buckets = 10`, `shard_size = 500` and `cache_limit = min(50 * shard_size, 50000)`.
```

Comment:
[This](https://github.com/elastic/elasticsearch/pull/57042) is going to be pretty good to have for this! Once you merge master into your PR you'll get it, and it makes this setting much less confusing, I think. As it stands now you I'm needing to crank the `search.max_buckets` up quite a bit or else I'll bump into it.

Reply:
Sorry, I'm not sure I understand. What will we get once I merge in master?

Reply:
Right now your aggregation loves to make a lot of buckets and then reduce them later. That PR will make us quite a bit more tollerant of that. It'll still be fairly expensive but that is ok.

Reply:
Ah, I see. I was reading this at night time and I had fl.ux on to remove blue colours, so I couldn't see that you linked the PR! Cool that makes sense.